### PR TITLE
Adding IAM Role ARN as a constraint for EC2 authentication

### PR DESCRIPTION
**Pull request by ifuyivara**
In the latest build 0.6.0 AMI ID was added as the first constraint to authenticate from an EC2 instance. This is really helpful for a lot of organizations that use different AWS AMI per application. Role tags was also added to allow organizations that use a base AMI to tag their instances for different role access in conjunction with an AMI ID. 

However our organization works a little bit different (or a mix of both). While we do have one base AMI this AMI is constantly changing (with security updates or added technologies) and will make it really hard to keep the Vault roles definitions in sync with the fast phased deployment cycles and AMI updates (think about tens of microservices, different apps, different technologies, each within a different autoscaling group, etc etc). After going back and forward with out DevOps and Security team we identified the IAM Role as the most effective way to achieve a secure identification and constraint for authentication. We assign those roles per application and is part of the autoscaling group and deployment process.

This adds IAM Role ARN as a constraint for authenticating against vault and obtaining a token for a given role. This can be used instead of the AMI ID but keeps both features (AMI ID has precedence if both are defined in the role definition).

To define a vault role with and IAM Role ARN as the constraint you can do the following:

<code>
vault write auth/aws-ec2/role/dev-role bound_iam_role_arn=arn:aws:iam::3233223121:instance-profile/vault-ec2 policies=prod,dev max_ttl=500h
</code>

Note we used the instance profile since this is what gets assigned to the EC2 instances. You can find this value for your role on the AWS console on IAM -> Roles

When you send the instance PKCS7 all the checks for the instance status will be the same and if the bound_iam_role_arn is set then it will verify that it matches the instance IAM Role ARN.

This will have the added benefit to work on Lambda functions since an AMI ID is not present but an IAM Role ARN is always set. 

Role tags will continue to work just the same as they work on AMI ID constraints. 

Thank you
-Ivan

### CODE REVIEW

**Comment by jhmartin** [comment]
IAM roles have the benefit of being immutable for a given instance -- a node may have privileges to update its tags, but there is no way for a subverted node to update the name of its IAM role.

File: builtin/credential/aws-ec2/path_login.go [hunk 1]
```diff
@@ -244,13 +244,21 @@ func (b *backend) pathLoginUpdate(
...
 	// AMI ID specified as a constraint on the role.
-	if identityDoc.AmiID != roleEntry.BoundAmiID {
+	if roleEntry.BoundAmiID != "" && identityDoc.AmiID != roleEntry.BoundAmiID {
 		return logical.ErrorResponse(fmt.Sprintf("AMI ID %s does not belong to role %s", identityDoc.AmiID, roleName)), nil
 	}
 
+	// Check if the IAM Role ARN of the instance trying to login matches the
+	// IAM Role ARN specified as a constraint on the role.
+	iamRoleArn := ""
+	iamRoleArn = *instanceDesc.Reservations[0].Instances[0].IamInstanceProfile.Arn
```
Comment (vishalnayak) [inline builtin/credential/aws-ec2/path_login.go]: `Reservations` and `Instances` will be checked to be non-nil in the `validateInstance` method, but `IamInstanceProfile` is not. Please have a `nil` check for that here.

Also for `Arn`, since that is a pointer too.

**Comment by vishalnayak** [comment]
@ifuyivara Please add or update a test, and update the docs.
Just FYI, `bound_account_id` is being added here: https://github.com/hashicorp/vault/pull/1523

**Comment by ifuyivara** [comment]
Thanks @vishalnayak I added the missing validation. I also added the tests and documentation and tried to make it so it will be easier to merge with #1523 but let me know if it's ok.

File: builtin/credential/aws-ec2/backend_test.go [hunk 2]
```diff
@@ -1126,6 +1131,7 @@ func TestBackendAcc_LoginAndWhitelistIdentity(t *testing.T) {
 		"policies":     "root",
 		"max_ttl":      "120s",
 		"bound_ami_id": "wrong_ami_id",
+		"bound_iam_role_arn" : "wront_iam_role_arn",
```
Comment (vishalnayak) [inline builtin/credential/aws-ec2/backend_test.go]: The login attempt with this data will fail as expected, not because of `wrong_iam_role_arm`, but because of `wrong_ami_id`.

After the login request with this data is complete, role needs to be updated again with correct AMI ID and wrong ARN, and a login request should be initiated again, with an expected failure.

Let me know if this makes sense to you.

Comment (ifuyivara) [inline builtin/credential/aws-ec2/backend_test.go]: It does. I missed that. I'll fix it in a minute. Thanks!

**Comment by vishalnayak** [comment]
Thanks for submitting this!

File: builtin/credential/aws-ec2/path_login.go [hunk 3]
```diff
@@ -244,13 +244,29 @@ func (b *backend) pathLoginUpdate(
...
-	if identityDoc.AmiID != roleEntry.BoundAmiID {
+	if roleEntry.BoundAmiID != "" && identityDoc.AmiID != roleEntry.BoundAmiID {
 		return logical.ErrorResponse(fmt.Sprintf("AMI ID %s does not belong to role %s", identityDoc.AmiID, roleName)), nil
 	}
 
+	// Check if the IAM Role ARN of the instance trying to login matches the
+	// IAM Role ARN specified as a constraint on the role.
+	if roleEntry.BoundIamARN != "" {
+		if instanceDesc.Reservations[0].Instances[0].IamInstanceProfile == nil {
+			return nil, fmt.Errorf("Iam instance profile in instance description is nil")
```
Comment (jefferai) [inline builtin/credential/aws-ec2/path_login.go]: pedantic: s/iam/IAM/

File: builtin/credential/aws-ec2/path_login.go [hunk 4]
```diff
@@ -244,13 +244,29 @@ func (b *backend) pathLoginUpdate(
...
+		}
+		if instanceDesc.Reservations[0].Instances[0].IamInstanceProfile.Arn == nil {
+			return nil, fmt.Errorf("ARN in instance description is nil")
+		}
+		iamRoleArn := ""
```
Comment (jefferai) [inline builtin/credential/aws-ec2/path_login.go]: What's the need for this? Just curious...

File: builtin/credential/aws-ec2/path_role.go [hunk 5]
```diff
@@ -211,10 +216,18 @@ func (b *backend) pathRoleCreateUpdate(
...
 
+	boundIamARNStr, ok := data.GetOk("bound_iam_role_arn")
+	if ok {
+		roleEntry.BoundIamARN = boundIamARNStr.(string)
+	}
+
 	// At least one bound parameter should be set. Currently, only
-	// 'bound_ami_id' is supported. Check if that is set.
+	// 'bound_ami_id' and 'bound_iam_role_arn' are supported. Check if one of them is set.
 	if roleEntry.BoundAmiID == "" {
```
Comment (jefferai) [inline builtin/credential/aws-ec2/path_role.go]: I'd make this a switch, since there are going to be more of these...

Comment (vishalnayak) [inline builtin/credential/aws-ec2/path_role.go]: Yes. `switch` would be more readable. Changed in #1523.

File: builtin/credential/aws-ec2/path_login.go (hunk 4 above)
Comment (vishalnayak) [inline builtin/credential/aws-ec2/path_login.go]: I don't think it is necessary. Changed in #1523.

File: builtin/credential/aws-ec2/path_login.go (hunk 3 above)
Comment (vishalnayak) [inline builtin/credential/aws-ec2/path_login.go]: Jeff, I think the author wrote `Iam` because the document returned by the AWS API had the name `IamInstanceProfile` in it.

`bound_account_id` is being added in #1523. There I have made this `IamInstanceProfile` there.

**Comment by dcosson** [comment]
Glad to see this get merged! I've been looking forward to ec2 integration but mapping AMI to role doesn't fit well into my setup either, plus I had trouble understanding the rather nuanced timing/replay attacks that the docs go into detail about preventing. Using the immutable Iam role and accepting that any user on the instance can always authenticate against the matching Vault role is nice, and similar to the way that Iam roles authenticate you against other AWS services.

Nice work!

**Comment by jefferai** [comment]
Unless I'm missing something, we aren't verifying the role against the AWS API, only the supplied metadata doc. This leaves this method wide open to replay attacks.

**Comment by jefferai** [comment]
Never mind on the earlier comment; missed that we were checking against the API data not the instance data. However, will this still work for lambda or is it a different endpoint than fetching EC2 info?

**Comment by ifuyivara** [comment]
I'm having a hard time getting the PKCS7 signature from a lambda function (it's probably not possible). So it won't be able to use the same endpoint. I'm investigating the best way to get a similar signature from Lambda and figure out a way to make it work with the same auth backend. It will probably end up it's own auth backend all together

**Comment by jefferai** [comment]
@ifuyivara I actually talked to someone from AWS recently who didn't have any helpful advice on identifying Lambda functions. It's not just about getting signed metadata -- I think he said all Lambda functions run with the same ARN (e.g. same instance ID) so you can't do anything like TOFU. The only way to have a realistic answer, I think, would be to figure out a way to verify that an auth call came from *a* Lambda instance, such that you could then take this info and verify against the IAM role.

**Comment by dcosson** [comment]
@jefferai what's the replay attack concern and need for the TOFU pattern? Am I right in thinking it's only a concern if you want the instance's access to change over time, i.e. the instance should have access to the vault role at first but then at some later time you don't want it to have access anymore, in case an attacker gets onto the instance and gets the signed metadata?

It seems like if I want the same behavior as IAM instance profile access to AWS services (where any user on the instance always has access to the Vault role but at least doesn't have access to any other roles) I can use the most lenient settings and trust any valid signature not just TOFU. Or am I missing something?

**Comment by jefferai** [comment]
> It seems like if I want the same behavior as IAM instance profile access to AWS services (where any user on the instance always has access to the Vault role but at least doesn't have access to any other roles) I can use the most lenient settings and trust any valid signature not just TOFU. Or am I missing something?

Yes  :-)  What you're missing is that most users _don't_ actually want any given user on the instance to be able to access Vault. The purpose of TOFU is to have a single trusted service make the initial authentication call, *without* allowing other services/processes/users on the instance able to then access Vault. The trusted service can then either pass the Vault token along/make it available on the FS, or can fetch required secrets and make them available as appropriate to users/services on the box.

To put it another way, many of our users consider IAM information to be far too easy to intercept by anyone that gets access to a box. Rather than have Vault have the same issues, the Vault method allows you to have better control over authentication information, while still allowing you to share it with all users if desired.

**Comment by dcosson** [comment]
Makes sense, thanks for the clarification 👍

**Comment by ktheory** [comment]
I'm excited to use this feature too.

Will it be part of the 6.1 release?

**Comment by jefferai** [comment]
Yes, as it's been merged!
